File: api/infrastructure/dto/stamp.go
/*
EmployeeのDTO
*/
package dto

import (
	"time"

	entity "github.com/Tomoya185-miyawaki/attend-log-gin/entity/stamp"
	"github.com/Tomoya185-miyawaki/attend-log-gin/helper"
	"gorm.io/gorm"
)

type Stamp struct {
	ID             uint        `gorm:"primary_key"`
	EmployeeID     uint        `gorm:"index;not null"`
	Status         StampStatus `gorm:"not null"`
	StampStartDate time.Time   `gorm:"not null"`
	StampEndDate   time.Time
	DeletedAt      gorm.DeletedAt `gorm:"default:null"`
}

type Stamps []Stamp

type StampCount struct {
	Cnt int
}

type StampStatus uint8

const (
	Attend = StampStatus(1)
	Rest   = StampStatus(2)
)

type StampRequestStatus int

const (
	AttendStart = StampRequestStatus(1)
	RestStart   = StampRequestStatus(2)
	RestEnd     = StampRequestStatus(3)
	WorkingEnd  = StampRequestStatus(4)
)

func (stamp Stamp) ConvertToModel() *entity.Stamp {
	if !stamp.StampEndDate.IsZero() {
		return &entity.Stamp{
			ID:             entity.ID(stamp.ID),
			EmployeeID:     entity.EmployeeID(stamp.EmployeeID),
			Status:         entity.Status(stamp.Status),
			StampStartDate: entity.StampStartDate(helper.TimeToString(stamp.StampStartDate)),
			StampStart:     entity.StampStart(helper.TimeToStringDuration(stamp.StampStartDate)),
			StampEndDate:   entity.StampEndDate(helper.TimeToString(stamp.StampEndDate)),
			StampEnd:       entity.StampEnd(helper.TimeToStringDuration(stamp.StampEndDate)),
		}
	}
	return &entity.Stamp{
		ID:             entity.ID(stamp.ID),
		EmployeeID:     entity.EmployeeID(stamp.EmployeeID),
		Status:         entity.Status(stamp.Status),
		StampStartDate: entity.StampStartDate(helper.TimeToString(stamp.StampStartDate)),
		StampStart:     entity.StampStart(helper.TimeToStringDuration(stamp.StampStartDate)),
	}
}

func (stamps Stamps) ConvertToModel() *entity.Stamps {
	result := make(entity.Stamps, len(stamps))

	for idx, stamp := range stamps {
		var stampEntity entity.Stamp
		if !stamp.StampEndDate.IsZero() {
			stampEntity = entity.Stamp{
				ID:             entity.ID(stamp.ID),
				EmployeeID:     entity.EmployeeID(stamp.EmployeeID),
				Status:         entity.Status(stamp.Status),
				StampStartDate: entity.StampStartDate(helper.TimeToString(stamp.StampStartDate)),
				StampStart:     entity.StampStart(helper.TimeToStringDuration(stamp.StampStartDate)),
				StampEndDate:   entity.StampEndDate(helper.TimeToString(stamp.StampEndDate)),
				StampEnd:       entity.StampEnd(helper.TimeToString(stamp.StampEndDate)),
			}
		} else {
			stampEntity = entity.Stamp{
				ID:             entity.ID(stamp.ID),
				EmployeeID:     entity.EmployeeID(stamp.EmployeeID),
				Status:         entity.Status(stamp.Status),
				StampStartDate: entity.StampStartDate(helper.TimeToString(stamp.StampStartDate)),
				StampStart:     entity.StampStart(helper.TimeToStringDuration(stamp.StampStartDate)),
			}
		}
		result[idx] = stampEntity
	}

	return &result
}

File: front/src/util/stamp.ts
import { StampsByEmployeeId } from '@/types/stamp'
import { StampStatus, StampAttendRestStatus } from '@/enums/stamp'

/**
 * 出勤状況を表す情報を返す
 * @return {array} 出勤状況を情報テキスト（1番目は表示するテキスト、2番目はボタンのテキスト、3番目は登録するstatus）
 */
export function getAttendRestStatus(stamps: StampsByEmployeeId[]) {
  let text = '出勤前'
  let buttonText = '出勤'
  let status = StampStatus.Attend
  if (stamps.length > 0) {
    stamps.sort((a, b) => a.status - b.status)
    for (let i = 0; i < stamps.length; i++) {
      if (stamps[i].status === StampAttendRestStatus.AttendLeaving && stamps[i].stamp_end_date !== '') {
        text = '退勤済み'
        buttonText = ''
        return [text, buttonText, status]
      }
      // 出勤済みの場合（後続処理を優先するためreturnしない）
      if (stamps[i].status === StampAttendRestStatus.AttendLeaving && stamps[i].stamp_start_date !== '') {
        text = '出勤済み'
        buttonText = '休憩開始'
        status = StampStatus.RestStart
      }
      // 休憩終了している場合
      if (stamps[i].status === StampAttendRestStatus.Rest && stamps[i].stamp_end_date !== '') {
        text = '休憩終了済み'
        buttonText = '退勤'
        status = StampStatus.Leaving
        return [text, buttonText, status]
      }
      // 休憩開始している場合
      if (stamps[i].status === StampAttendRestStatus.Rest && stamps[i].stamp_start_date !== '') {
        text = '休憩開始済み'
        buttonText = '休憩終了'
        status = StampStatus.RestEnd
        return [text, buttonText, status]
      }
    }
  }
  return [text, buttonText, status]
}